File: ruby/day03_part1.rb
# frozen_string_literal: true

lines = File.read('input/day03.txt').split("\n")

numbers_included = {}
data = {}

lines.each.with_index do |line, i|
  data[i] = {}
  numbers_included[i] = []
  nums = line.to_enum(:scan, /\d+/)
             .map { Regexp.last_match }
             .map { |m| [(m.offset(0).first..m.offset(0).last).to_a, m[0].to_i] }.to_h
  data[i][:nums] = nums
  joiners = line.to_enum(:scan, /[^.\d]/)
                .map { Regexp.last_match }
                .map { |m| m.offset(0) }
                .flatten
  data[i][:joiners] = joiners
  nums.each { |is, n| numbers_included[i] << n unless is.intersection(joiners).empty? }
  if i.positive?
    data[i - 1][:nums].each { |is, n| numbers_included[i - 1] << n unless is.intersection(joiners).empty? }
    nums.each { |is, n| numbers_included[i] << n unless is.intersection(data[i - 1][:joiners]).empty? }
  end
end

p numbers_included.values.flatten.reduce(:+)

File: src/bin/day03_part1.rs
use regex::Regex;
use std::{fs, ops::Range};

fn main() {
    let number_re = Regex::new(r"\d+").unwrap();
    let joiner_re = Regex::new(r"[^.\d]").unwrap();

    let mut numbers_per_line = Vec::<Vec<(i32, Range<usize>)>>::new();
    let mut joiners_per_line = Vec::<Vec<usize>>::new();
    let mut numbers_included: Vec<i32> = Vec::new();

    let input = fs::read_to_string("input/day03.txt").unwrap();
    let lines = input.split('\n');

    for (line_index, line) in lines.enumerate() {
        let nums: Vec<(i32, Range<usize>)> = number_re
            .find_iter(line)
            .map(|m| (m.as_str().parse().unwrap(), m.start()..m.end()))
            .collect();
        let joiners: Vec<usize> = joiner_re
            .find_iter(line)
            .flat_map(|m| Vec::from_iter((m.start() - 1)..(m.end() + 1)))
            .collect();

        for (n, n_range) in &nums {
            if joiners.iter().any(|j| n_range.contains(j)) {
                numbers_included.push(*n);
            }
            if line_index > 0
                && joiners_per_line[line_index - 1]
                    .iter()
                    .any(|j| n_range.contains(j))
            {
                numbers_included.push(*n);
            }
        }

        if line_index > 0 {
            for (n, n_range) in &numbers_per_line[line_index - 1] {
                if joiners.iter().any(|j| n_range.contains(j)) {
                    numbers_included.push(*n);
                }
            }
        }

        numbers_per_line.push(nums);
        joiners_per_line.push(joiners);
    }

    dbg!(numbers_included.iter().sum::<i32>());
}
